## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
column_names = df.columns.tolist()
print("Index", column_names)

Index ['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id', 'order_date', 'order_week', 'order_year', 'item_id', 'category', 'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty', 'line_number']


In [3]:
# Use the describe function to gather some basic statistics
df_statistics = df.describe()
df_statistics

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# See the next 7 cells below

In [5]:
# Single value
value = df.loc[0, 'first']
print(value)

Donald


In [6]:
# Single row
row = df.loc[0]
print(row)

first                                    Donald
last                                    Harding
job                                Immunologist
phone                        793-904-7725x39308
email          harding.donald.7185@sullivan.com
client_id                                 58515
order_id                                8953482
order_date                           2023-04-28
order_week                                   17
order_year                                 2023
item_id                          EUD29711-63-6U
category                                  decor
subcategory                            wall art
unit_price                               1096.8
unit_cost                                762.71
unit_weight                                 7.5
qty                                         105
line_number                                   1
Name: 0, dtype: object


In [7]:
# Single column
column = df['last']
print(column)

0          Harding
1            Myers
2           Watson
3            Baker
4         Schwartz
           ...    
54634        Reese
54635        Moore
54636    Gutierrez
54637      Solomon
54638         Bell
Name: last, Length: 54639, dtype: object


In [8]:
# A subset of multiple rows and columns
subset = df.loc[0:4, ['job', 'phone']]
subset

job               phone
0           Immunologist  793-904-7725x39308
1        Music therapist    201.442.4543x942
2           Immunologist   687.737.9424x8503
3  Accounting technician    827-788-8123x012
4           Chiropractor        265-829-3643

In [9]:
# Sort by unit price column
sorted_df = df.sort_values(by='unit_price')
sorted_df

first        last                                  job  \
19201     Jorge       Myers       Chartered certified accountant   
51194     David       Brady                      Publishing copy   
30966    Rachel      Morton                 Brewing technologist   
23070   Melissa    Thompson        Telecommunications researcher   
42302  Michelle       Smith                     Therapist, drama   
...         ...         ...                                  ...   
9938       Leah   Middleton               Geologist, engineering   
12724   Derrick       Moore                             Musician   
27224     Sarah  Richardson  Exhibitions officer, museum/gallery   
16156     James   Wilkinson          Conservator, museum/gallery   
7558     Steven     Perkins          Engineer, building services   

                       phone                              email  client_id  \
19201       514.586.1248x433              myers.jorge@smith.com      84432   
51194             6597485343              dbrady@cunningham.com      69260   
30966          (575)733-9673  r.morton.1758@martinez-barber.com      55217   
23070  +1-504-380-8552x27062            mthompson8773@brown.biz      49451   
42302    (822)251-2751x51533           smith.m.5154@patrick.org      68370   
...                      ...                                ...        ...   
9938           (928)705-9665      middleton.l@harris-turner.biz      34878   
12724           358.661.5483       derrick.moore.2602@pope.info      41908   
27224             2056834802    richardsonsmile4695@merritt.com      79361   
16156       991.297.5872x313   jamesperson@gonzalez-hoffman.biz      17697   
7558            997-280-8748          perkinss@carter-hayes.com      36023   

       order_id  order_date  order_week  order_year         item_id  \
19201   1527738  2023-01-28           4        2023  ZAZ05162-35-5X   
51194   9770746  2023-02-20           8        2023  ZAZ05162-35-5X   
30966   5445711  2023-01-18           3        2023  ZAZ05162-35-5X   
23070   8531050  2023-01-14           2        2023  ZAZ05162-35-5X   
42302   1398752  2023-01-24           4        2023  ZAZ05162-35-5X   
...         ...         ...         ...         ...             ...   
9938    4951455  2023-04-20          16        2023  BBE61778-67-4A   
12724   6034977  2023-05-20          20        2023  BBE61778-67-4A   
27224   7487007  2023-05-06          18        2023  BBE61778-67-4A   
16156   4923598  2023-03-05           9        2023  BBE61778-67-4A   
7558    7637317  2023-01-19           3        2023  BBE61778-67-4A   

          category        subcategory  unit_price  unit_cost  unit_weight  \
19201        decor             plants        0.01       0.01         8.04   
51194        decor             plants        0.01       0.01         8.04   
30966        decor             plants        0.01       0.01         8.04   
23070        decor             plants        0.01       0.01         8.04   
42302        decor             plants        0.01       0.01         8.04   
...            ...                ...         ...        ...          ...   
9938   consumables  bathroom supplies     1396.23     846.27         9.53   
12724  consumables  bathroom supplies     1396.23     846.27         9.53   
27224  consumables  bathroom supplies     1396.23     846.27         9.53   
16156  consumables  bathroom supplies     1396.23     846.27         9.53   
7558   consumables  bathroom supplies     1396.23     846.27         9.53   

       qty  line_number  
19201  151            8  
51194  180            5  
30966   44            3  
23070    6            1  
42302  152            4  
...    ...          ...  
9938     9            3  
12724   59            0  
27224   87            4  
16156  213            7  
7558   298            6  

[54639 rows x 18 columns]

In [10]:
# Filter clients that purchased a quantity greater than or equal to 1,000
filtered_df = df[df['qty'] >= 1000]
filtered_df

first      last                                 job  \
24        Glenn    Garcia                    Textile designer   
25        Peter   Roberts          Architectural technologist   
79        Erica     Colon                     Learning mentor   
98     Nicholas  Williams                      Radio producer   
119      Joshua     Garza                  Physicist, medical   
...         ...       ...                                 ...   
54597     Kevin    Obrien                      Microbiologist   
54601     Julie    Holmes  Museum/gallery exhibitions officer   
54617    Jerome  Erickson        Administrator, Civil Service   
54619     Laura     Owens                   Buyer, industrial   
54638    Dwayne      Bell             Chief Financial Officer   

                       phone                                 email  client_id  \
24              573.981.6374     determinegarcia5685@velasquez.com      33804   
25                9238354425     roberts.peter@warren-sandoval.com      85413   
79          580-422-4977x362            colon.erica.1035@reese.biz      31268   
98         393.560.7337x2039     nicholas.williams.6317@strong.biz      56590   
119    +1-893-589-7899x11282                 storyjoshua@floyd.com      40490   
...                      ...                                   ...        ...   
54597             7112448766          line.kevin@hart-williams.com      39939   
54601          (818)928-3636           tenholmes@maddox-thomas.com      80508   
54617  001-434-476-9810x1983  erickson.cause@matthews-robinson.biz      36531   
54619           753-596-3400              lauraowens6281@baker.org      27771   
54638      606-619-7667x9508                      belld@galvan.com      24007   

       order_id  order_date  order_week  order_year         item_id  \
24      6024379  2023-01-26           4        2023  XZZ76042-95-6X   
25      9362283  2023-01-18           3        2023  BAD54487-56-3C   
79      4582776  2023-04-09          14        2023  DCM29568-31-1X   
98      7777768  2023-01-15           2        2023  UZX16759-11-7E   
119     5422023  2023-04-21          16        2023  XBZ40996-40-6A   
...         ...         ...         ...         ...             ...   
54597   6430193  2023-04-07          14        2023  UZX16759-11-7E   
54601   8767718  2023-02-27           9        2023  BDU54519-00-1M   
54617   5593997  2023-05-21          20        2023  ZEA01443-18-5A   
54619   8016714  2023-05-13          19        2023  DAA84430-62-7E   
54638   7489403  2023-01-04           1        2023  ZCZ05589-50-5D   

          category         subcategory  unit_price  unit_cost  unit_weight  \
24     consumables      paper products        5.17       3.53         6.90   
25     consumables      paper products        7.14       5.10         1.23   
79     consumables   bathroom supplies      169.54     133.28         2.20   
98        software          accounting       13.68      10.71         7.95   
119      furniture              stands      342.71     291.02        14.76   
...            ...                 ...         ...        ...          ...   
54597     software          accounting       13.68      10.71         7.95   
54601    furniture              stands      149.56     120.37         7.93   
54617    furniture              stands        6.86       4.06         6.89   
54619  consumables   bathroom supplies        4.67       2.39        10.49   
54638     software  project management        3.01       1.61         2.07   

          qty  line_number  
24       2163            2  
25     129944            4  
79      29073            0  
98       1167            4  
119      4896            5  
...       ...          ...  
54597    4445            0  
54601    1302            3  
54617    3291            5  
54619    2319            5  
54638    1031            5  

[2711 rows x 18 columns]

In [11]:
# Get unique values in the category column
unique_categories = df['category'].unique()

# Print unique values
print("Types of item categories:", unique_categories)

# Get the number of unique values in the category column
num_unique_categories = df['category'].nunique()

# Print the number of unique item categories
print("Number of item categories:", num_unique_categories)

Types of item categories: ['decor' 'consumables' 'software' 'furniture' 'electronics']
Number of item categories: 5


In [12]:
# What three item categories had the most entries?

# Count the occurrences of each item category
category_counts = df['category'].value_counts()

# Display the top 3 categories with the most entries
top_3_categories = category_counts.head(3)
print(top_3_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [13]:
# For the category with the most entries,
# which subcategory had the most entries?

# Filter the DataFrame to include only rows where the category is 'consumables'
consumables_df = df[df['category'] == 'consumables']

# Count the occurrences of each subcategory within the 'consumables' category
subcategory_counts = consumables_df['subcategory'].value_counts()

# Find the subcategory with the most entries
top_subcategory = subcategory_counts.idxmax()

# Get the count of the most frequent subcategory
top_subcategory_count = subcategory_counts.max()

# Print the subcategory with the most entries and its count
print(f"{top_subcategory}    {top_subcategory_count}")

bathroom supplies    6424


In [14]:
# Which five clients had the most entries in the data?

# Count the occurrences of each client_id
client_counts = df['client_id'].value_counts()

# Display the top 5 clients with the most entries
top_5_clients = client_counts.head(5)
print(top_5_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [15]:
# Store the client ids of those top 5 clients in a list.

# Count the occurrences of each client_id
client_counts = df['client_id'].value_counts()

# Get the top 5 client IDs
top_5_client_ids = client_counts.head(5).index.tolist()

print(top_5_client_ids)

[33615, 66037, 46820, 38378, 24741]


In [16]:
# How many total units (the qty column) did the
# client with the most entries order order?

# Identify the client ID with the most entries
most_entries_client_id = df['client_id'].value_counts().idxmax()

# Filter the DataFrame to include only rows for this client
client_df = df[df['client_id'] == most_entries_client_id]

# Sum the 'qty' column for these rows
total_units_ordered = client_df['qty'].sum()

print(total_units_ordered)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [17]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

# Create a new column 'line_subtotal' by multiplying 'unit_price' and 'qty'
df['line_subtotal'] = df['unit_price'] * df['qty']

# Print only the specified columns without the index
df[['unit_price', 'qty', 'line_subtotal']].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [18]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

# Calculate total weight for each row
df['total_weight'] = df['unit_weight'] * df['qty']

# Define conditions for shipping price
conditions = [
    (df['total_weight'] > 50),
    (df['total_weight'] <= 50)
]

# Define corresponding shipping prices
choices = [
    df['total_weight'] * 7,
    df['total_weight'] * 10
]

# Create the 'shipping_price' column based on conditions
df['shipping_price'] = np.select(conditions, choices)

# Print the DataFrame to see the new column
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [19]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

# Calculate the total before tax
total_before_tax = df['line_subtotal'] + df['shipping_price']

# Calculate the tax amount as 9.25% of the total before tax
tax_amount = total_before_tax * 0.0925

# Calculate the total price by adding the total before tax and the tax amount
df['line_price'] = total_before_tax + tax_amount

# Round the line price to the nearest penny
df['line_price'] = df['line_price'].round(2)

# Print the DataFrame to see the new column
df[['line_subtotal', 'shipping_price', 'line_price']].head(3)

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [20]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

# Calculate the cost for each line
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

# Print the DataFrame to see the new column
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [21]:
# Create a column for the profit of
# each line using line cost and line price

# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']

# Display the DataFrame to see the new column along with other relevant information
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [22]:
# Check your work using the totals above

# Group by 'order_id' and sum 'line_price' to get the total price for each order
order_totals = df.groupby('order_id')['line_price'].sum()

# Create a dictionary of the expected totals for easy comparison
expected_totals = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Check the calculated totals against the expected totals
for order_id, expected_total in expected_totals.items():
    calculated_total = order_totals.get(order_id, 0)
    print(f"Order {order_id} total: ${calculated_total:.2f}")

Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [23]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

# Identify the top 5 clients by quantity
client_counts = df.groupby('client_id').size().sort_values(ascending=False)
top_5_clients = client_counts.head(5).index.tolist()

# Calculate the total spent by each of the top 5 clients
client_spending = df.groupby('client_id')['line_price'].sum()
top_clients_spending = client_spending.loc[top_5_clients]

# Print the results
for client_id, total_spent in top_clients_spending.items():
    print(f"{client_id}: ${total_spent:.2f}")

33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [24]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Identify the top 5 clients by quantity
top_5_clients = client_counts.head(5).index.tolist()

# Calculate the totals for each of the top 5 clients
client_totals = df.groupby('client_id')[['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']].sum()
top_clients_totals = client_totals.loc[top_5_clients]

# Sort by total profit
top_clients_totals_sorted = top_clients_totals.sort_values('line_profit', ascending=False)

# Display the results
top_clients_totals_sorted

qty  shipping_price   line_price    line_cost  line_profit
client_id                                                               
24741      239862      5126448.37  82268892.02  45688899.71  36579992.31
38378       73667      3429455.40  12906550.87   9634720.98   3271829.89
66037       43018      1395151.85  10259514.79   7004482.98   3255031.81
46820       75768      1601448.84   9743794.36   7007191.64   2736602.72
33615       64313      1828984.89   8377308.52   6175313.91   2201994.61

In [25]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# Rename the columns
top_clients_totals_sorted.rename(columns={
    'client_id': 'Client ID',
    'qty': 'Units',
    'shipping_price': 'Shipping',
    'line_price': 'Total Revenue',
    'line_cost': 'Total Cost',
    'line_profit': 'Total Profit'
}, inplace=True)

# Convert to millions of dollars
currency_columns = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']
for col in currency_columns:
    top_clients_totals_sorted[col] = top_clients_totals_sorted[col] / 1_000_000

# Format the currency columns
for col in currency_columns:
    top_clients_totals_sorted[col] = top_clients_totals_sorted[col].map('${:,.2f}M'.format)
    
# Display the results
top_clients_totals_sorted

Units Shipping Total Revenue Total Cost Total Profit
client_id                                                       
24741      239862   $5.13M       $82.27M    $45.69M      $36.58M
38378       73667   $3.43M       $12.91M     $9.63M       $3.27M
66037       43018   $1.40M       $10.26M     $7.00M       $3.26M
46820       75768   $1.60M        $9.74M     $7.01M       $2.74M
33615       64313   $1.83M        $8.38M     $6.18M       $2.20M

In [26]:
# Sort the updated data by "Total Profit" form highest to lowest
sorted_df = top_clients_totals_sorted.sort_values(by='Total Profit', ascending=False)

# Display the sorted DataFrame
sorted_df

Units Shipping Total Revenue Total Cost Total Profit
client_id                                                       
24741      239862   $5.13M       $82.27M    $45.69M      $36.58M
38378       73667   $3.43M       $12.91M     $9.63M       $3.27M
66037       43018   $1.40M       $10.26M     $7.00M       $3.26M
46820       75768   $1.60M        $9.74M     $7.01M       $2.74M
33615       64313   $1.83M        $8.38M     $6.18M       $2.20M

# Findings

An outlier of profit was achieved from client 24741. They purchased 239,862 items and provided the company with 82.27 million dollars in gross revenue, which cost the company 45.69 million dollars, for a total profit of 36.58 million dollars. The 4 other largest money making clients brought in a total of 11.47 million dollars in profit. This means that the biggest client brought in about 3.19 times the profits of the 2nd through 5th biggest clients combined.